source/common/upstream/ring_hash_lb.h

```diff
@@ -45,26 +45,29 @@ class RingHashLoadBalancer : public LoadBalancerBase,
   struct Ring {
     Ring(const Optional<envoy::api::v2::Cluster::RingHashLbConfig>& config,
          const std::vector<HostSharedPtr>& hosts);
-    HostConstSharedPtr chooseHost(LoadBalancerContext* context,
-                                  Runtime::RandomGenerator& random) const;
+    HostConstSharedPtr chooseHost(uint64_t hash) const;
 
     std::vector<RingEntry> ring_;
   };
-
   typedef std::shared_ptr<const Ring> RingConstSharedPtr;
 
+  struct PerPriorityState {
+    RingConstSharedPtr current_ring_;
+    bool global_panic_{};
+  };
+  typedef std::unique_ptr<PerPriorityState> PerPriorityStatePtr;
+
   struct LoadBalancerImpl : public LoadBalancer {
-    LoadBalancerImpl(ClusterStats& stats, Runtime::RandomGenerator& random,
-                     const RingConstSharedPtr& ring, bool global_panic)
-        : stats_(stats), random_(random), ring_(ring), global_panic_(global_panic) {}
+    LoadBalancerImpl(ClusterStats& stats, Runtime::RandomGenerator& random)
+        : stats_(stats), random_(random) {}
 
     // Upstream::LoadBalancer
     HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
 
     ClusterStats& stats_;
     Runtime::RandomGenerator& random_;
-    const RingConstSharedPtr ring_;
-    const bool global_panic_;
+    std::vector<PerPriorityStatePtr> per_priority_state_;
```

Comment:
Instead of copying the vectors can we just use shared_ptr to constant vector? This will make lock hold time smaller also. (I think per other comments the locking is not quite right unless I am missing something).

Reply:
I can but it's a perf hit for everything else since they then can't do in-place edits.
If we're worried about lock time how about a shared pointer for the ring hash where we refresh, which can be shared by all the threads?

Reply:
nit: I would either do shared_ptr for both per_priority_state_ and per_priority_load_, or for neither. It's fine with me either way (doing it out of lock with swap is fine). Any reason not to be consistent?